griptape/tools/google_drive/tool.py

```diff
@@ -0,0 +1,153 @@
+            ): str,
+            Literal(
+                "file_name",
+                description="name to give the file on Google Drive"
+            ): str,
+            Literal(
+                "mime_type",
+                description="MIME type of the file being uploaded"
+            ): str
+        })
+    })
+    def upload_file(self, params: dict) -> InfoArtifact | ErrorArtifact:
+        from googleapiclient.http import MediaFileUpload
+
+        values = params['values']
+
+        try:
+            service = self._build_client(self.UPLOAD_FILE_SCOPES, values)
+            file_metadata = {'name': values['file_name']}
+            media = MediaFileUpload(values['file_path'], mimetype=values['mime_type'])
+            file = service.files().create(body=file_metadata, media_body=media, fields='id').execute()
+
+            return InfoArtifact(f'A file was successfully uploaded with ID {file.get("id")}.')
+
+        except Exception as e:
+            logging.error(e)
+            return ErrorArtifact(f"error uploading file to Drive: {e}")
+
+    @activity(config={
+        "description": "Can be used to download a file from Google Drive",
+        "schema": Schema({
+            Literal(
+                "drive_owner_email",
+                description="email of the Google Drive's owner"
+            ): str,
+            Literal(
+                "file_id",
+                description="ID of the file to be downloaded from Google Drive"
+            ): str
+        })
+    })
+    def download_file(self, params: dict) -> TextArtifact | ErrorArtifact:
+
+        values = params["values"]
+
+        try:
+            service = self._build_client(self.LIST_FILES_SCOPES, values)
+            request = service.files().get_media(fileId=values["file_id"])
+            downloaded_file = request.execute()
+
+            return TextArtifact(downloaded_file)
+
+        except HttpError as e:
+            logging.error(e)
+            return ErrorArtifact(f"error downloading file from Drive: {e}")
+        except MalformedError:
+            logging.error("MalformedError occurred")
+            return ErrorArtifact("error downloading file due to malformed credentials")
+
+    @activity(config={
+        "description": "Can be used to search for a file in Google Drive",
+        "schema": Schema({
+            Literal(
+                "drive_owner_email",
+                description="email of the Google Drive's owner"
+            ): str,
+            Literal(
+                "file_name",
+                description="name of the file to search for in Google Drive"
+            ): str
+        })
+    })
+    def search_file(self, params: dict) -> ListArtifact | ErrorArtifact:
+
+        values = params["values"]
+
+        try:
+            service = self._build_client(self.LIST_FILES_SCOPES, values)
+            query = f"name='{values['file_name']}'"
+            results = service.files().list(q=query).execute()
```

Comment:
What kind of search does this do? Based on file name? Content? Metadata? All of the above? Depending on the answer we should re-structure the schema.

Reply:
based on file name

Reply:
Looks like it [supports](https://developers.google.com/drive/api/guides/search-files) multiple modes. I think it would be very useful to provide search based on the content of the file as well. Imagine asking the LLM "find all files in folder/foo/bar that contain 'vasily' in body."

We can move that to one of the activity schema params, so that the LLM can decide what type of search to perform.